config/src/main/java/com/yahoo/config/subscription/impl/ConfigSubscription.java

```diff
@@ -129,50 +154,54 @@ public boolean equals(Object o) {
         return false;
     }
 
-    void setConfigChanged(boolean changed) {
-        this.configChanged = changed;
-    }
-
-    void setGenerationChanged(boolean genChanged) {
-        this.generationChanged = genChanged;
-    }
 
     /**
      * Called from {@link ConfigSubscriber} when the changed status of this config is propagated to the clients
      */
-    public void resetChangedFlags() {
-        setConfigChanged(false);
-        setGenerationChanged(false);
+    public boolean isConfigChangedAndReset(Long requiredGen) {
+        ConfigState<T> prev = config.get();
+        while ( prev.getGeneration().equals(requiredGen) && !config.compareAndSet(prev, prev.createUnchanged())) {
+            prev = config.get();
+        }
+        // A false positive is a lot better than a false negative
+        return !prev.getGeneration().equals(requiredGen) || prev.isConfigChanged();
     }
 
-    public boolean isConfigChanged() {
-        return configChanged;
+    void setConfig(Long generation, T config) {
+        this.config.set(new ConfigState<>(true, generation, true, config));
     }
 
-    public boolean isGenerationChanged() {
-        return generationChanged;
+    // Only used by {@link FileConfigSubscription}
+    protected void setConfigIncGen(T config) {
+        ConfigState<T> prev = this.config.get();
+        setConfig(prev.getGeneration() + 1, config);
     }
 
-    void setConfig(T config) {
-        this.config = config;
+    // Only used by {@link FileConfigSubscription} and {@link ConfigSetSubscription}
+    protected void setConfigIfChangedIncGen(T config) {
+        ConfigState<T> prev = this.config.get();
+        this.config.set(new ConfigState<>(true, prev.getGeneration() + 1,
+                                          !config.equals(prev.getConfig()), config));
+    }
+    protected void setConfigIfChanged(T config) {
+        ConfigState<T> prev = this.config.get();
+        this.config.set(new ConfigState<>(true, prev.getGeneration(),
```

Comment:
when is this useful? it seems pretty self-conflicting; the config itself may have changed, the generation has not changed, but is tagged as changed...